### PR TITLE
Use rustc generated shims instead of interpreter hacks

### LINKED ISSUE #119: bad substs for tuple-like constructor function pointer
**Source**

```rust
fn main() {
    let f: fn(i32) -> Option<i32> = Some::<i32>;
    f(42);
}
```

**Output and error**

```
TRACE:miri::step:  _1 = std::option::Option<i32>::Some as fn(i32) -> std::option::Option<i32> (ReifyFnPointer)
TRACE:miri::eval_context:  _1: Undef
DEBUG:miri::memory: creating fn ptr: 2
TRACE:miri::eval_context:  _1: Ptr(Pointer { alloc_id: AllocId(2), offset: 0 })
TRACE:miri::memory:  Alloc 2:     function pointer: std::prelude::v1::Some: fn(i32) -> std::option::Option<i32>
TRACE:miri::step:  StorageLive(_3)
TRACE:miri::step:  _3 = _1
TRACE:miri::eval_context:  _3: Undef
TRACE:miri::eval_context:  _1: Ptr(Pointer { alloc_id: AllocId(2), offset: 0 })
TRACE:miri::memory:  Alloc 2:     function pointer: std::prelude::v1::Some: fn(i32) -> std::option::Option<i32>
TRACE:miri::eval_context:  _3: Ptr(Pointer { alloc_id: AllocId(2), offset: 0 })
TRACE:miri::memory:  Alloc 2:     function pointer: std::prelude::v1::Some: fn(i32) -> std::option::Option<i32>
TRACE:miri::step:  _2 = _3(const 42i32) -> bb1
TRACE:miri::eval_context:  _2: Undef
TRACE:miri::eval_context:  _3: Ptr(Pointer { alloc_id: AllocId(2), offset: 0 })
TRACE:miri::memory:  Alloc 2:     function pointer: std::prelude::v1::Some: fn(i32) -> std::option::Option<i32>
DEBUG:miri::memory: reading fn ptr: 2
error: internal compiler error: /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/librustc/ty/subst.rs:476: Type parameter `T/#0` (T/0) out of range when substituting (root type=Some(std::option::Option<T>)) substs=[]
```

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #119

~~Unfortunately I broke some things:~~

~~1904 success, 197 no mir, 206 crate not found, 71 failed, 46 C fn, 0 ABI, 121 unsupported, 6 intrinsic~~
~~1910 success, 197 no mir, 206 crate not found, 64 failed, 46 C fn, 0 ABI, 122 unsupported, 6 intrinsic~~
~~1918 success, 197 no mir, 206 crate not found, 56 failed, 46 C fn, 0 ABI, 122 unsupported, 6 intrinsic~~
~~1921 success, 199 no mir, 206 crate not found, 51 failed, 46 C fn, 0 ABI, 122 unsupported, 6 intrinsic~~
1925 success, 200 no mir, 206 crate not found, 46 failed, 46 C fn, 0 ABI, 122 unsupported, 6 intrinsic

```
  14 main function not found
  12 a raw memory access tried to access part of a pointer value as raw bytes
   9 the evaluated program panicked
   2 attempted to do math or a comparison on pointers into different allocations
   2 Overflow(Add) at /checkout/src/libcore/ops.rs:263:45: 263:57
   1 unused import: `std::option`
   1 tried to call a function with sig extern "C" fn(usize) -> Foo through a function pointer of type extern "C" fn(usize) -> u32
   1 tried to call a function with sig extern "C" fn(&isize) -> std::option::Option<&isize> through a function pointer of type extern "C" fn(&isize) -> &isize
   1 tried to access memory with alignment 1, but alignment 8 is required
   1 src/step.rs:250: static def id doesn't point to item
   1 internal compiler 
   1 Overflow(Neg) at /checkout/src/libcore/ops.rs:758:43: 758:45
```

### CODE REVIEW

**Comment by oli-obk** [comment]
And travis broke due to travis moving to rustup :tada:

File: src/terminator/intrinsic.rs [hunk 1]
```diff
@@ -299,35 +268,35 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
-                let ty = substs.type_at(0);
+                let ty = instance.substs.type_at(0);
                 let env = self.tcx.empty_parameter_environment();
                 let needs_drop = self.tcx.type_needs_drop_given_env(ty, &env);
                 self.write_primval(dest, PrimVal::from_bool(needs_drop), dest_ty)?;
             }
 
             "offset" => {
-                let pointee_ty = substs.type_at(0);
+                let pointee_ty = instance.substs.type_at(0);
```
Comment (eddyb) [inline src/terminator/intrinsic.rs]: I feel like all of these wouldn't need to be changed if you copied `substs` out. It's just a reference anyway.

Comment (oli-obk) [inline src/terminator/intrinsic.rs]: Yea I want to do some cleanups after everything works.

**Comment by dwrensha** [comment]
For the sake of quick cross-reference, here's a link to the copied-in code from `rustc_trans::monomorphize`: [monomorphize.rs](https://github.com/rust-lang/rust/blob/8dfc25e05e995c92a6093718198cca4bc2c85b63/src/librustc_trans/monomorphize.rs#L30-L277)

**Comment by oli-obk** [comment]
Ok... I fixed all the fallout.

We have one new `Overflow(Add)` and 4 ICEs (probably the ABI stuff)

@arielb1's changes fixed all 3 of our `Type parameter `T/#0` (T/0) out of range ` bugs

File: src/terminator/intrinsic.rs (hunk 1 above)
Comment (solson) [inline src/terminator/intrinsic.rs]: Do you still plan to do some cleanups? Should I wait until those are done before reviewing?

Comment (oli-obk) [inline src/terminator/intrinsic.rs]: Not in this PR. I wanted to separate it from any cleanup noise

Comment (eddyb) [inline src/terminator/intrinsic.rs]: IMO it's a very simple change in this function that cleans up the diff.

Comment (oli-obk) [inline src/terminator/intrinsic.rs]: done

**Comment by solson** [comment]
Sorry for the delay. I've added @eddyb and @oli-obk as collaborators on the repo so you two can make progress without blocking on me. Unfortunately, I don't have much free time/energy these days to spend on this. That said, I'll still be following the progress and reviewing as much as I can. I'll also be looking for ways to fit this back into my life again, but I shouldn't make any promises.

Feel free to either commit to master or continue doing PRs - I'll try to keep pace either way. Just don't consider my approval a requirement from now on.

**Comment by eddyb** [comment]
@oli-obk Can you update to the latest nightly or should I try?

**Comment by eddyb** [comment]
>  Unable to update https://github.com/quininer/byteorder.git?branch=i128#ef51df29

Probably needs to be switched to 9bab6d7783f81da50feb234a120c918d9eabba6e now, or just the branch left in, if that's possible (it's a feature branch so they force pushed).

**Comment by oli-obk** [comment]
Won't get to this before tuesday

**Comment by oli-obk** [comment]
You need to change the got path to the official master branch. We are on an unofficial repo right now

**Comment by eddyb** [comment]
@oli-obk Did the i128 branch get merged or something? AFAICT miri needs the i128 support. So I updated the hash to the top of the branch (hope you don't mind me committing on your master - you should use a different branch because @solson and me get to modify your master branch :))).

**Comment by eddyb** [comment]
@oli-obk `tests/compile-fail/never_say_never.rs` looks like it now enters an infinite loop?

**Pull request merged**